File: neetcode/python/853.car-fleet.py
#
# @lc app=leetcode id=853 lang=python3
#
# [853] Car Fleet
#

# @lc code=start
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        position_speed = [
            (position[i], speed[i])
            for i in range(len(speed))    
        ]

        position_speed = sorted(
            position_speed,
            key=lambda pair: pair[0],
            reverse=True
        )

        stack = []
        for pos, speed in position_speed:
            time_taken = (target - pos) / speed
            stack.append(time_taken)

            n = len(stack)
            if n > 1 and stack[n-1] <= stack[n-2]:
                stack.pop()
        return len(stack)
        
# @lc code=end

File: neetcode/go/853.car-fleet.go
/*
 * @lc app=leetcode id=853 lang=golang
 *
 * [853] Car Fleet
 */

// @lc code=start
func carFleet(target int, position []int, speed []int) int {
	positionSpeedMap := make(map[int]int, len(position))
	for i := 0; i < len(position); i++ {
		positionSpeedMap[position[i]] = speed[i]
	}

	sort.Slice(position, func(i, j int) bool {
		return position[i] > position[j]
	})

	var stack = []float64{}

	for _, pos := range position {
		timeTaken := float64(target-pos) / float64(positionSpeedMap[pos])
		n := len(stack)
		if n == 0 || stack[n-1] < timeTaken {
			stack = append(stack, timeTaken)
		}
	}
	return len(stack)
}

// @lc code=end